File: Chapter 03/C#/Section_3_3_4_ISP/Before/Midfielder.cs
using System;

namespace Chapter03.ISP.Before;

/// <summary>
/// The Polluted Implementation (The Victim of the ISP Violation).
/// 
/// ARCHITECTURE PROBLEM: Because the Midfielder wants to participate in 
/// the ITrainingSession, it is FORCED by the compiler to implement methods 
/// it has no business knowing about. 
/// 
/// To make the compiler happy, the developer has to write "dummy" methods 
/// or throw NotImplementedExceptions. This creates "code rot" and sets traps 
/// for other developers who might accidentally call these methods expecting 
/// them to actually do something.
/// </summary>
public class Midfielder : ITrainingSession
{
    /// <summary>
    /// A valid method for this specific class.
    /// </summary>
    public void PracticeShooting() 
    { 
        Console.WriteLine("  [Midfielder] Practicing shooting drills."); 
    }
    
    /// <summary>
    /// A valid method for this specific class.
    /// </summary>
    public void PracticeTackling() 
    { 
        Console.WriteLine("  [Midfielder] Practicing slide tackles."); 
    }
    
    /// <summary>
    /// 🚨 ARCHITECTURE WARNING: The Midfielder doesn't need this, but the 
    /// Fat Interface demands it! We are forced to throw an exception.
    /// </summary>
    public void PracticeDivingSaves() 
    {
        throw new NotImplementedException("Midfielders don't play in the net!");
    }
    
    /// <summary>
    /// 🚨 ARCHITECTURE WARNING: Another useless method forced upon us.
    /// </summary>
    public void PracticeHandDistribution() 
    {
        throw new NotImplementedException("Midfielders can't use their hands!");
    }
}

File: Chapter 03/Java/src/main/java/com/grokkingsoftwarearchitecture/chapter03/section_3_3_4_isp/before/Midfielder.java
package com.grokkingsoftwarearchitecture.chapter03.section_3_3_4_isp.before;

/**
 * The Polluted Implementation (The Victim of the ISP Violation).
 * * ARCHITECTURE PROBLEM: Because the Midfielder wants to participate in 
 * the TrainingSession, it is FORCED by the compiler to implement methods 
 * it has no business knowing about. 
 * * To make the compiler happy, the developer has to write "dummy" methods 
 * or throw UnsupportedOperationExceptions. This creates "code rot" and sets traps 
 * for other developers who might accidentally call these methods expecting 
 * them to actually do something.
 */
public class Midfielder implements TrainingSession {
    
    /**
     * A valid method for this specific class.
     */
    public void practiceShooting() {
        System.out.println("  [Midfielder] Practicing shooting drills.");
    }

    /**
     * A valid method for this specific class.
     */
    public void practiceTackling() {
        System.out.println("  [Midfielder] Practicing slide tackles.");
    }

    /**
     * 🚨 ARCHITECTURE WARNING: The Midfielder doesn't need this, but the 
     * Fat Interface demands it! We are forced to throw an exception.
     */
    public void practiceDivingSaves() {
        throw new UnsupportedOperationException("Midfielders don't play in the net!");
    }

    /**
     * 🚨 ARCHITECTURE WARNING: Another useless method forced upon us.
     */
    public void practiceHandDistribution() {
        throw new UnsupportedOperationException("Midfielders can't use their hands!");
    }
}